Add Damerau-Levenshtein support for calculating string distance

reblws opened the pull request:
Issue: #333 

Hi, this PR adds Damerau-Levenshtein as an option to the `LevenshteinDistance` function. This makes the function consider transposition as an edit operation, so:

```
Regular Levenshtein:
'az' vs. 'za' = 2
``
Damerau-Levenshtein:
'az' vs. 'za' = 1 
```

This PR adds 2 new options for the `LevenshteinDistance` function:
```
damerau: boolean - whether we want to add transposition as a valid edit operation
transposition_cost: number - cost of a transpose operation
```
In terms of space-complexity, this PR requires an additional O(length of source word) space. We could optimize this space cost by choosing the shorter of the target and source strings to store, but this might affect the results if the search option is enabled, so I chose not to do this.

syzer commented on README.md:
````diff
@@ -130,6 +130,28 @@ Output:
+Full Damerau-Levenshtein matching can be used if you want to consider character transpositions as a valid edit operation.
+
+```javascript
+console.log(natural.LevenshteinDistance("az", "za", { damerau: true }));
````
would not be better to use ```natural.damarauDistance(...)``` ?

kkoch986 replied:
yea i agree with @syzer seems like it would be easier to use that way. @reblws would you be able to quickly update that?

reblws replied:
Yeah, no problem. I'll put this in a separate file, along with an extra option for simpler Optimal String Alignment.

I feel a bit uneasy about pasting functions, namely `getMinCostSubstring` and `_getMatchStart` from `levenshtein_distance.js`. I'll just copy & paste them for now, but let me know if you want them to be in a common module.

Also, last thing: should the function be called `DamerauLevenshteinDistance` or `DamerauDistance`?

kkoch986 replied:
hmm yea sorry i forgot how this was all structured, we could change the way it exports snd have the one file export both, then update [this](https://github.com/NaturalNode/natural/blob/master/lib/natural/index.js#L78) to expose both. does that make sense?

reblws replied:
Ok, so we'll export both `DamerauLevenshteinDistance` and `LevenshteinDistance` from `levenshtein_distance.js` and then update `natural/index.js` to import from `levenshtein_distance` explicitly?

reblws replied:
Ok, all done

reblws commented:
Hi, I changed the `levenshtein_distance.js` file to export two separate functions: `LevenshteinDistance` and `DamerauLevenshteinDistance`. I also added an option for simple Optimal String Alignment if `options.restricted = true`

syzer commented on spec/damerau_levenshtein_spec.js:
```diff
@@ -0,0 +1,63 @@
+/*
+Copyright (c) 2011, Chris Umbel
```
About copyright:
It can increase projects size considerably.

Cannot we just add Chris Umbel to readme as contributor?
what do you think @reblws ?

syzer replied:
and remove the copyright from that file ?

reblws replied:
My apologies, I didn't catch this when I was pasting these tests onto a new file. I removed it. 

I do think it would be nice if there was a single file where copyright notices went so they aren't all over the place.

syzer commented on spec/damerau_levenshtein_spec.js:
```diff
@@ -0,0 +1,63 @@
+var damerauLevenshtein = require('../lib/natural/distance/levenshtein_distance')
+  .DamerauLevenshteinDistance;
+
+describe('DamerauLevenshtein', function () {
```
👍

kkoch986 commented:
im ok with this, the copyright is all in the license too so that should cover everything. thanks for this!
